examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,410 @@
+// Copyright (c) 2017-2022, Lawrence Livermore National Security, LLC and other CEED contributors.
+// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.
+//
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This file is part of CEED:  http://github.com/ceed
+
+//                        libCEED + PETSc DMSwarm Example
+//
+// This example demonstrates a simple usage of libCEED with DMSwarm.
+// This example combines elements of PETSc src/impls/dm/swam/tutorials/ex1.c and src/impls/dm/swarm/tests/ex6.c
+//
+// Build with:
+//
+//     make dmswarm [PETSC_DIR=</path/to/petsc>] [CEED_DIR=</path/to/libceed>]
+//
+// Sample runs:
+//
+//     ./dmswarm
+
+/// @file
+/// libCEED example using PETSc with DMSwarm
+const char help[] = "libCEED example using PETSc with DMSwarm\n";
+
+#include <ceed.h>
+#include <petscdmplex.h>
+#include <petscdmswarm.h>
+#include <petscds.h>
+#include <petscfe.h>
+#include <petscksp.h>
+#include <petsc/private/petscfeimpl.h> /* For interpolation */
+
+#include "include/bpsproblemdata.h"
+#include "include/petscutils.h"
+
+static PetscScalar EvalU(PetscInt dim, const PetscScalar x[]) {
+  PetscScalar result = 1, center = 0.1;
+  for (PetscInt d = 0; d < dim; d++) {
+    result *= tanh(x[d] - center);
+    center += 0.1;
+  }
+  return result;
+}
+
+static PetscScalar EvaldU(PetscInt dim, const PetscScalar x[], PetscInt direction) {
+  PetscScalar result = 1, center = 0.1;
+  for (PetscInt d = 0; d < dim; d++) {
+    if (d == direction) result *= 1.0 / cosh(x[d] - center) / cosh(x[d] - center);
+    else result *= tanh(x[d] - center);
+    center += 0.1;
+  }
+  return result;
+}
+
+PetscErrorCode DMSwarmCreateReferenceCoordinates(DM dm_swarm, DM dm_mesh, IS *is_points, Vec *ref_coords) {
```

Comment:
Note - I put `dm_mesh` as an argument here because there is no `DMSwarmGetCellDM`. This is fine since this will be an internal only method, but I think we should *also* add `DMSwarmGetCellDM` since that will be useful for us.

Reply:
Isn't this `DMSwarmGetCellDM`? https://gitlab.com/petsc/petsc/-/blob/zach/bug-dm-swarm-tensor-permutation/src/dm/impls/swarm/swarm.c#L872

Reply:
dunno how I missed that